velox/functions/sparksql/URLFunctions.h

```diff
@@ -0,0 +1,310 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/regex.hpp>
+#include <folly/CPortability.h>
+#include <cctype>
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+/// Performs initial validation of the URI.
+/// Checks if the URI contains ascii whitespaces or
+/// unescaped '%' chars.
+bool isValidURI(StringView input) {
+  const char* p = input.data();
+  const char* end = p + input.size();
+  char buf[3];
+  buf[2] = '\0';
+  char* endptr;
+  for (; p < end; ++p) {
+    if (facebook::velox::functions::stringImpl::isAsciiWhiteSpace(*p)) {
+      return false;
+    }
+
+    if (*p == '%') {
+      if (p + 2 < end) {
+        buf[0] = p[1];
+        buf[1] = p[2];
+        strtol(buf, &endptr, 16);
+        p += 2;
+        if (endptr != buf + 2) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+static const boost::regex kUriRegex(
+    "(([a-zA-Z][a-zA-Z0-9+.-]*):)?" // scheme:
+    "([^?#]*)" // authority and path
+    "(?:\\?([^#]*))?" // ?query
+    "(?:#(.*))?"); // #fragment
+template <typename T>
+T submatch(const boost::cmatch& match, int indx);
+
+template <>
+FOLLY_ALWAYS_INLINE std::string_view submatch(
+    const boost::cmatch& match,
+    int idx) {
+  const auto& sub = match[idx];
+  return std::string_view(sub.first, sub.length());
+}
+
+template <>
+FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
+  const auto& sub = match[idx];
+  return StringView(sub.first, sub.length());
+}
+
+template <typename TInString>
+bool parse(const TInString& rawUrl, boost::cmatch& match) {
+  if (!isValidURI(rawUrl)) {
+    return false;
+  }
+  return boost::regex_match(
+      rawUrl.data(), rawUrl.data() + rawUrl.size(), match, kUriRegex);
+}
+// PARSE_URL(url, partToExtract) → string
+// PARSE_URL(url, partToExtract, key) → string
+//
+// Extracts a part of a URL. The partToExtract argument can be one of
+// 'PROTOCOL', 'HOST', 'PATH', 'REF', 'AUTHORITY', 'FILE', 'USERINFO',
+// 'QUERY', 'FRAGMENT'.
+//
+// If the URL is invalid, return null.
+template <typename T>
+struct ParseUrlFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  static constexpr int kAuthPath = 3;
+  static constexpr int kQuery = 4;
+  static constexpr int kHost = 3;
+  static constexpr int kProto = 2;
+  static constexpr int kRef = 5;
+  // submatch indexes for authorityMatch
+  static constexpr int kPathHasAuth = 2;
+  static constexpr int kPathNoAuth = 3;
+  static constexpr int kUser = 1;
+  static constexpr int kPass = 2;
+  static constexpr int kPort = 4;
+
+  FOLLY_ALWAYS_INLINE bool call(
```

Comment:
This method is very long. Let's not inline it and let's break it up into smaller pieces.

Reply:
Created private helper functions for path, file, userinfo, and authority